src/renderer/base/renderer.cpp

```diff
@@ -664,6 +664,71 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
     }
 }
 
+void Renderer::_PaintBufferLineOutputHelper(_In_ IRenderEngine* const pEngine,
+                                            const ROW& row,
+                                            const COORD target)
+{
+    size_t cols = 0;
+    const ATTR_ROW& attrRow = row.GetAttrRow();
+    const std::vector<Cluster> rowClusters = row.GetClusters();
+
+    const size_t runCount = attrRow.GetNumberOfRuns();
+    size_t runIndex = 0;
+    // And hold the point where we should start drawing.
+    auto screenPoint = target;
+
+    size_t runStart = 0;
+    // This outer loop will continue until we reach the end of the text we are trying to draw.
+    while (runIndex < runCount)
+    {
+        const TextAttributeRun run = attrRow.GetRunByIndex(runIndex);
+        const TextAttribute attr = run.GetAttributes();
+        const size_t runLength = run.GetLength();
+
+        const auto currentRunColor = attr;
+
+        // Update the drawing brushes with our color.
+        THROW_IF_FAILED(_UpdateDrawingBrushes(pEngine, currentRunColor, false));
+
+        // Advance the point by however many columns we've just outputted and reset the accumulator.
+        screenPoint.X += gsl::narrow<SHORT>(cols);
+        cols = 0;
+
+        size_t count = 0;
+        std::vector<Cluster> clusters;
+        clusters.reserve(runLength);

```

Comment:
The `reserve` here is intended to reduce vector reallocation and it works great. Without `reserve`, the performance is heavily penalized.